### ISSUE TITLE
Chromium via ozone/wayland can not play videos

### ISSUE THREAD

**Issue by maxhille** [NixOS/nixpkgs#104885]
**Describe the bug**

Chromium on ozone/wayland can not play videos. Has errors in the log

**To Reproduce**
1. be on `unstable` with sway
2. start via `$ chromium -enable-features=UseOzonePlatform -ozone-platform=wayland`
3. try to open https://www.youtube.com/watch?v=hVvEISFw9w0

**Expected behavior**
Video should play

**Observed behavior**
Video does not play. Errors in the log which correlate to trying to watch videos

**Additional context**
console log:
```
~ $ chromium -enable-features=UseOzonePlatform -ozone-platform=wayland
[8990:8990:1125/115341.050188:ERROR:sandbox_linux.cc(374)] InitializeSandbox() called with multiple threads in process gpu-process.
[8957:9022:1125/115341.079959:ERROR:object_proxy.cc(621)] Failed to call method: org.freedesktop.DBus.Properties.Get: object_path= /org/freedesktop/UPower: org.freedesktop.DBus.Error.ServiceUnknown: The name org.freedesktop.UPower was not provided by any .service files
[8957:9022:1125/115341.084162:ERROR:object_proxy.cc(621)] Failed to call method: org.freedesktop.UPower.GetDisplayDevice: object_path= /org/freedesktop/UPower: org.freedesktop.DBus.Error.ServiceUnknown: The name org.freedesktop.UPower was not provided by any .service files
[8957:9022:1125/115341.085450:ERROR:object_proxy.cc(621)] Failed to call method: org.freedesktop.UPower.EnumerateDevices: object_path= /org/freedesktop/UPower: org.freedesktop.DBus.Error.ServiceUnknown: The name org.freedesktop.UPower was not provided by any .service files
WARNING: Kernel has no file descriptor comparison support: Function not implemented
[8990:8990:1125/115341.173663:ERROR:shared_context_state.cc(74)] Skia shader compilation error
------------------------

Errors:

[8990:8990:1125/115341.182664:ERROR:shared_context_state.cc(74)] Skia shader compilation error
------------------------

Errors:

[8990:8990:1125/115341.875115:ERROR:shared_context_state.cc(74)] Skia shader compilation error
------------------------

Errors:

[8990:8990:1125/115341.877472:ERROR:shared_context_state.cc(74)] Skia shader compilation error
------------------------

Errors:

[8990:8990:1125/115341.881079:ERROR:shared_context_state.cc(74)] Skia shader compilation error
------------------------

Errors:

[8990:8990:1125/115341.902295:ERROR:shared_context_state.cc(74)] Skia shader compilation error
------------------------

Errors:

[8990:8990:1125/115341.946859:ERROR:shared_context_state.cc(74)] Skia shader compilation error
------------------------

Errors:

[8990:8990:1125/115342.459109:ERROR:shared_context_state.cc(74)] Skia shader compilation error
------------------------

Errors:

[9017:13:1125/115347.527278:ERROR:client_native_pixmap_dmabuf.cc(252)] Failed to mmap dmabuf: Permission denied (13)
[8990:8990:1125/115347.534829:ERROR:shared_context_state.cc(74)] Skia shader compilation error
------------------------

Errors:

[8990:8990:1125/115347.673750:ERROR:shared_context_state.cc(74)] Skia shader compilation error
------------------------

Errors:

[8990:8990:1125/115347.680073:ERROR:shared_context_state.cc(74)] Skia shader compilation error
------------------------

Errors:

[9017:13:1125/115352.144819:ERROR:client_native_pixmap_dmabuf.cc(252)] Failed to mmap dmabuf: Permission denied (13)
```

**Notify maintainers**
@primeos 

**Metadata**
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.
```
~ $ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.9.9, NixOS, 21.03pre253635.2247d824fe0 (Okapi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.9`
 - channels(mh): `""`
 - channels(root): `"nixos-21.03pre253635.2247d824fe0"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

~ $ chromium --version
Chromium 87.0.4280.66
```

**Comment by primeos** [comment]
I'm aware of this issue but I'm not sure if this is due to our packaging or an upstream issue.
Do you know if playing videos is supposed to work with both `use_x11` and `use_ozone` (e.g. if it works with Google Chrome or on other distributions)?

I know that at least VA-API isn't implemented for Ozone/Wayland yet but software playback is probably supposed to work (though I don't know anything about that and Ozone/Wayland is still experimental). 

The following error message is potentially interesting in this regard:
```
[9017:13:1125/115347.527278:ERROR:client_native_pixmap_dmabuf.cc(252)] Failed to mmap dmabuf: Permission denied (13)
```

(IIRC I got a "not implemented" error from `client_native_pixmap_dmabuf.cc` when I tested it.)

**Comment by maxhille** [comment]
At least when I run `chromium -enable-features=UseOzonePlatform` it starts in XWayland and videos work fine. I don't know much more about the whole thing, sorry.

Also toggling `chrome://flags/#enable-accelerated-video-decode` does not seem to change anything (console log shows the same errors).

**Comment by primeos** [comment]
Let's close this then as there's a high chance that this is an upstream "bug". Additionally Ozone and VA-API are both experimental.

AFAIK [ozone/linux: add --enable-features=UseOzonePlatform](https://bugs.chromium.org/p/chromium/issues/detail?id=1085700) caused a few regressions (e.g. [1119303](https://bugs.chromium.org/p/chromium/issues/detail?id=1119303)) and the video playback issue could be one of them (or it is e.g. due to incompatible system dependencies, Wayland compositor, etc.). In either case this will likely resolve itself in the future (M88+) and I'm keeping an eye on that stuff anyway.

**Comment by primeos** [comment]
Ok, big update (had some time and gave it another shot): I found [ozone/wayland: do not return supported buffer formats if gbm disabled.](https://chromium.googlesource.com/chromium/src/+/3fd9f2846e7ac74bde9c13e9641ab0ae55ae7636) which seemed a likely candidate (we use `ozone_platform_gbm` to make the build work with `use_system_minigbm`) but that commit is already part of 87.0.4280.66. However, `chrome://gpu/` didn't report all GpuMemoryBuffers as software only so I gave `--disable-gpu-memory-buffer-video-frames` a shot which fixed video playback on my system (YAY!).

So launching Chromium with `--disable-gpu-memory-buffer-video-frames` should "fix" video playback in the meantime (not sure how high the performance penalty is but watching 1080p on my system was fine).

**Comment by MetaDark** [comment]
Huh, I wonder if the issue is related to the graphics card / drivers you're using. I'm also using `nixos-21.03pre253635.2247d824fe0` and the video plays perfectly at 2160p60. I'm using a Radeon RX 590 with Mesa 20.2.2.

This is the only output I'm getting from chromium:
```
amdgpu: os_same_file_description couldn't determine if two DRM fds reference the same file description.
If they do, bad things may happen!
[9160:9160:1125/114224.038224:ERROR:sandbox_linux.cc(374)] InitializeSandbox() called with multiple threads in process gpu-process.
```

**Comment by primeos** [comment]
> Huh, I wonder if the issue is related to the graphics card / drivers you're using.

Yeah, I was wondering about that too. The drivers are a very likely candidate for such issues (there are a lot of driver specific hacks/workarounds in Chromium, their supported features differ, and GPU drivers have bugs/regressions too).

From [ozone/wayland: do not return supported buffer formats if gbm disabled.](https://chromium.googlesource.com/chromium/src/+/3fd9f2846e7ac74bde9c13e9641ab0ae55ae7636) it seems less driver related (if ozone_platform_gbm/use_wayland_gbm = false still disables the dmabuf path) but if the driver doesn't support any of Chromium's GpuMemoryBuffers (at least the ones used for playing the video) it should coincidentally work.

Anyway, thanks for your additional information! :)
Could you please also provide the "GpuMemoryBuffers Status" output from `chrome://gpu/` (screenshot or text)?

I forgot to safe it on my test system (some where supported) but on my laptop they're currently all software only (but I think that isn't due to the driver):
```
GpuMemoryBuffers Status
R_8	Software only
R_16	Software only
RG_88	Software only
BGR_565	Software only
RGBA_4444	Software only
RGBX_8888	Software only
RGBA_8888	Software only
BGRX_8888	Software only
BGRA_1010102	Software only
RGBA_1010102	Software only
BGRA_8888	Software only
RGBA_F16	Software only
YVU_420	Software only
YUV_420_BIPLANAR	Software only
P010	Software only
```

**Comment by MetaDark** [comment]
> Could you please also provide the "GpuMemoryBuffers Status" output from chrome://gpu/ (screenshot or text)?

Here's the output I get under `GpuMemoryBuffers Status`:
| Pixel Format | Status
-- | --
R_8 | CAMERA_AND_CPU_READ_WRITE, SCANOUT_CPU_READ_WRITE
R_16 | Software only
RG_88 | Software only
BGR_565 | GPU_READ
RGBA_4444 | Software only
RGBX_8888 | GPU_READ, SCANOUT, SCANOUT_CPU_READ_WRITE
RGBA_8888 | GPU_READ, SCANOUT, SCANOUT_CPU_READ_WRITE
BGRX_8888 | GPU_READ, SCANOUT, SCANOUT_CPU_READ_WRITE
BGRA_1010102 | SCANOUT, SCANOUT_CPU_READ_WRITE
RGBA_1010102 | SCANOUT, SCANOUT_CPU_READ_WRITE
BGRA_8888 | GPU_READ, SCANOUT, SCANOUT_CPU_READ_WRITE
RGBA_F16 | Software only
YVU_420 | GPU_READ, SCANOUT_VEA_READ_CAMERA_AND_CPU_READ_WRITE
YUV_420_BIPLANAR | SCANOUT_CAMERA_READ_WRITE, SCANOUT_CPU_READ_WRITE, SCANOUT_VEA_READ_CAMERA_AND_CPU_READ_WRITE
P010 | Software only

**Comment by primeos** [comment]
I've tried to fix this in 2d5bb88418f7082136de44a5b1665efc8dd3ad80 but didn't realize that this was a no-op as `ozone_platform_gbm` is now called `ozone_platform_drm` (better name) and defaults to `false`.
I also tried M88 but still couldn't play any videos without the extra flag.

@MetaDark I've compared your `GpuMemoryBuffers Status` output to the one on my test system but surprisingly I get exactly the same output. Are the `gfx::BufferFormats` also all unsupported on your system (not sure how they're different but I recently saw them being used in the VA-API code - not sure about the normal video decoding tough)?:
```
gfx::BufferFormats supported for allocation and texturing	R_8: not supported, R_16: not supported, RG_88: not supported, BGR_565: not supported, RGBA_4444: not supported, RGBX_8888: not supported, RGBA_8888: not supported, BGRX_8888: not supported, BGRA_1010102: not supported, RGBA_1010102: not supported, BGRA_8888: not supported, RGBA_F16: not supported, YVU_420: not supported, YUV_420_BIPLANAR: not supported, P010: not supported
```

**Comment by MetaDark** [comment]
@primeos Yep, they're all unsupported:

```
gfx::BufferFormats supported for allocation and texturing	R_8: not supported, R_16: not supported, RG_88: not supported, BGR_565: not supported, RGBA_4444: not supported, RGBX_8888: not supported, RGBA_8888: not supported, BGRX_8888: not supported, BGRA_1010102: not supported, RGBA_1010102: not supported, BGRA_8888: not supported, RGBA_F16: not supported, YVU_420: not supported, YUV_420_BIPLANAR: not supported, P010: not supported
```

**Comment by zachlatta** [comment]
In case it's worth adding, video playback on Chromium in Wayland works fine for me when using the `amdgpu` video drivers.

On my laptop that's using the internal Intel GPU, videos won't play and I get the errors others are mentioning above unless I run with `--disable-gpu-memory-buffer-video-frames`. If I use that flag, video playback works.

**Comment by primeos** [comment]
I've just tried removing `--disable-gpu-memory-buffer-video-frames` again and surprisingly playing videos does work now on my laptop with an Intel GPU (and Mesa's `iris` driver). It also works when I explicitly specify `--enable-gpu-memory-buffer-video-frames` so this might be fixed now (at least for Intel users, I guess - and some other setups already worked before, like in the comment above).

I'm a bit surprised that it already works because I thought https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/10851 would fix it (while I'm still on Mesa 20.0.1).
Anyway, https://bugs.chromium.org/p/chromium/issues/detail?id=1095578 should be fixed now (Mesa issue: https://gitlab.freedesktop.org/mesa/mesa/-/issues/3119).
